### PR TITLE
Add CompositeEndpointGroup

### LINKED ISSUE #1897: Add CompositeEndpointGroup that aggregates multiple endpoint groups
I believe we're missing a simple way to register multiple endpoint groups as one - this would mostly be used when given a list of hosts that may resolve to multiple A records, each would be its own DnsAddressEndpointGroup but we'd like to round-robin over all the records. We currently have `OrElseEndpointGroup` that picks one or the other, but not one that iterates through all of them.

### PR DESCRIPTION
**Pull request by anuraaga**
Found while working on https://github.com/openzipkin/zipkin/blob/81a481b490399a797bc5d04c00ee7de400a4f015/zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java#L527

Also makes `Endpoint` implement `EndpointGroup` as a group of itself so APIs can accept both groups and non-groups transparently.

Fixes #1897

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.util.AbstractListenable;
+
+/**
+ * An {@link EndpointGroup} that merges the result of any number of other {@link EndpointGroup}s.
+ */
+public final class CompositeEndpointGroup extends AbstractListenable<List<Endpoint>> implements EndpointGroup {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java]: How about making this package-private and adding `EndpointGroup.of(EndpointGroup...)` and `of(Iterable<EndpointGroup>)`?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        for (EndpointGroup endpointGroup : endpointGroups) {
+            newEndpoints.addAll(endpointGroup.endpoints());
+        }
+
+        return merged = newEndpoints.build();
+    }
+
+    @Override
+    public void close() {
+        Closer closer = Closer.create();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java]: nit: `final`?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            return merged;
+        }
+
+        if (!dirtyUpdater.compareAndSet(this, 1, 0)) {
+            // Another thread might be updating merged at this time, but endpoint groups are allowed to take a
+            // little bit of time to reflect updates.
+            return merged;
+        }
+
+        ImmutableList.Builder<Endpoint> newEndpoints = ImmutableList.builder();
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java]: nit: `final`? +1

File: core/src/test/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroupTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+class CompositeEndpointGroupTest {
+
+    private static final Endpoint FOO = Endpoint.of("foo");
+    private static final Endpoint BAR = Endpoint.of("bar");
+    private static final Endpoint CAT = Endpoint.of("cat");
+    private static final Endpoint DOG = Endpoint.of("dog");
+
+    @Test
+    void normal() {
+        DynamicEndpointGroup group1 = new DynamicEndpointGroup();
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroupTest.java]: nit: final? +2

File: core/src/test/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroupTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        group1.setEndpoints(ImmutableList.of(FOO, BAR));
+        DynamicEndpointGroup group2 = new DynamicEndpointGroup();
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroupTest.java]: nit: final? +3

File: core/src/test/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroupTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        group2.setEndpoints(ImmutableList.of(CAT, DOG));
+
+        CompositeEndpointGroup composite = new CompositeEndpointGroup(group1, group2);
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroupTest.java]: nit: final? +4

File: core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java (hunk 1 above)
Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java]: Thanks - will make the change but putting this PR on hold until #1940

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java]: Merged #1940 :-)

**Comment by anuraaga** [comment]
Just pushed WIP commits (now I need my deleted method back ;) )

One feeling I had while tweaking this is it might be nice if `Endpoint` was `EndpointGroup`, similar to `EventLoop` and `EventLoopGroup`.

**Comment by trustin** [comment]
> One feeling I had while tweaking this is it might be nice if `Endpoint` was `EndpointGroup`, similar to `EventLoop` and `EventLoopGroup`. 

Indeed. I like the idea :-)

**Comment by anuraaga** [comment]
Do you mind taking a skim of https://github.com/line/armeria/pull/1939/commits/e276167b8fa993ac6c3d5fc8eda1cf7605040bf5 ?

There are goods and bads of making `Endpoint` `EndpointGroup`, especially when writing documentation. How does it look?

**Comment by anuraaga** [comment]
@trustin Just a reminder if you can give a quick skim of the general direction that'd be awesome

**Comment by adriancole** [comment]
chasing up things, would be nice to have this merged for whatever next release

**Comment by adriancole** [comment]
I don't know if this is the issue for it or not, but maybe we need to cache last exception so that if an endpoint group is empty we can know why then and there? here's an example https://github.com/openzipkin/zipkin/issues/2763#issuecomment-523006770

**Comment by trustin** [comment]
@anuraaga Sorry I'm late. It looks OK-ish to me. The methods related with the initial endpoints look somewhat unnatural, but I think it's not that bad. What do you think, @minwoox ?

**Comment by trustin** [comment]
@adriancole There's an issue for that: #1910

**Comment by anuraaga** [comment]
Sorry for the huge delay, ready for another look

File: core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java [hunk 7]
```diff
@@ -41,6 +46,50 @@ static EndpointGroup empty() {
...
+     * {@code endpointGroups} can be instances of {@link Endpoint} as well, any {@link EndpointGroup}s and
+     * {@link Endpoint} will all be combined into a single {@link EndpointGroup} that contains the total set.
+     */
+    static EndpointGroup of(Iterable<EndpointGroup> endpointGroups) {
+        requireNonNull(endpointGroups, "endpointGroups");
+
+        final List<EndpointGroup> groups = new ArrayList<>();
+        final List<Endpoint> staticEndpoints = new ArrayList<>();
+        for (EndpointGroup endpointGroup : endpointGroups) {
+            if (endpointGroup instanceof Endpoint) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java]: Should we also spread the endpoints in the `StaticEndpointGroup`?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/StaticEndpointGroup.java [hunk 8]
```diff
@@ -26,7 +26,10 @@
 
 /**
  * A static immutable {@link EndpointGroup}.
+ *
+ * @deprecated use {@link EndpointGroup}
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/StaticEndpointGroup.java]: - use -> Use
- missing period
- `{@link EndpointGroup#of(EndpointGroup...)}` ?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java (hunk 7 above)
Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java]: Sure also realized I was missing tests for these special cases

File: core/src/main/java/com/linecorp/armeria/client/endpoint/StaticEndpointGroup.java (hunk 8 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/StaticEndpointGroup.java]: I guess you forgot this :-)

Comment (anuraaga) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/StaticEndpointGroup.java]: Oops

File: core/src/test/java/com/linecorp/armeria/client/endpoint/EndpointGroupTest.java [hunk 9]
```diff
@@ -18,20 +18,101 @@
...
 
         assertThat(emptyEndpointGroup.orElse(endpointGroup2).endpoints())
                 .isEqualTo(endpointGroup2.endpoints());
         assertThat(endpointGroup1.orElse(endpointGroup2).endpoints())
                 .isEqualTo(endpointGroup1.endpoints());
     }
+
+    @Test
+    void normal() {
+        DynamicEndpointGroup group1 = new DynamicEndpointGroup();
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/endpoint/EndpointGroupTest.java]: nit: we could add `final` modifier for the variables used in this class

**Comment by minwoox** [comment]
Thanks, @anuraaga!

**Pull request merged**